Bugfix sed

pmelchior opened the pull request:
SED initialization code was not robust for negative/zero flux in some bands. As such, such an SED is not a problem. I therefore changed the behavior from raising an exception to logging a warning.

I also, moved the SED correction (for PSF width variations) from `get_pixel_sed` to the relevant `init_` functions because this is where it's needed. It's confusing otherwise.

fred3m commented on scarlet/source.py:
```diff
@@ -224,10 +252,32 @@ def init_multicomponent_source(sky_coord, frame, observation, bg_rms, flux_perce
         morphs[k] /= morphs[k].max()
+        if np.all(morphs[k] <= 0):
+            msg = "Zero or negative morphology for component {} at y={}, x={}"
+            logger.warning(msg.format(k, *skycoords))
```
Shouldn't this check be before you normalize `morphs[k]`? If all of the values are negative then dividing by the max will always give a positive result, so I'm not sure that this will ever be true.

pmelchior commented on scarlet/source.py:
```diff
@@ -36,17 +36,6 @@ def get_pixel_sed(sky_coord, observation):
     sed = observation.images[:, pixel[0], pixel[1]].copy()
-    if observation.frame.psfs is not None:
-        # Account for the PSF in the intensity
-        sed /= observation.frame.psfs.max(axis=(1, 2))
```
Because it does something that the user doesn't expect. It gives you the SED of a pixel *and* makes the first step of a two-step correction for PSF width changes. That is highly confusing. It's conceptually better to keep the SED rescaling to the init methods.

fred3m replied:
In that case it might still be better to do the rescaling in the function but introduce a `rescale_by_psf` parameter to optionally rescale by the PSF. That would help avoid duplicate code but still make it clear what is being done.

pmelchior replied:
I disagree. This entire operation only makes sense when you also do the multiplication with the blend frame PSF because it's in essence a matching operation. That's what I meant by two-step procedure. It's very unclear what this rescaling does if the two parts are performed in very different regions of the code.

fred3m replied:
I get your point, but in that case we should probably have a function like `_get_normalized_sed` that implements lines 108-117 as a function that `init_extended_source` and `init_combined_extended_source` can both call.

pmelchior replied:
new function `get_psf_sed` that does that.

pmelchior commented:
@fred3m , I've addressed all comments, and created a simpler implementation of MultiComponentSource that actually uses a common center for all components. You can merge if you are fine with it.

fred3m commented on scarlet/component.py:
```diff
@@ -363,8 +363,8 @@ def update(self):
-        for component in self.components:
-            component.update()
+        for node in self._tree:
+            node.update()
```
Good catch, looks like for a hierarchical tree we would have been calling update multiple times.

pmelchior replied:
I don't think so. We've called only the components, so nothing got done multiple times. The problem was that a node that is not a leaf (i.e. a multi-component source) never had its `update` method called.

fred3m replied:
Ah, good point. You're right, I was thinking that it called `update` for all of the components, but when a source with multiple components had its `update` method called then some of them would be called again. I didn't notice that only the leaves were called.

pmelchior commented on scarlet/source.py:
```diff
@@ -567,32 +567,65 @@ def __init__(self, frame, sky_coord, observation, bg_rms, thresh=1, flux_percent
+        # ensure adherence to constraints
+        self.update()
+
+    def update(self):
```
That was never my intention. Also see below.

In the case of a multi-component source, we have aspects of the update that need to be done consistently across all of its components, e.g. finding the center. So, the source, not the component, needs to have an aspect of the update. At that point it was easier to hollow out the component updates entirely and avoid having an internal class definition that becomes eventually exposed to the user.

fred3m replied:
Ok, that makes sense.
